### DISCUSSION TITLE
How to define optional parameters in effects in typescript ?

### DISCUSSION THREAD

**Discussion by StringKe** [rematch/rematch#924] [Q&A] [answered]
```typescript
const test = {
  effects: dispatch => ({
    select(id?:string, state):void {
      dispatch.testSelect(id?id:state.oldSelect);
    }
  })
}
```

**Comment by tianzhich** [comment, marked as answer]
I'm not sure what you are asking here. In your above example, you already defined an optional parameter `id?: string`.

**Reply by StringKe** [reply]
![image](https://user-images.githubusercontent.com/31089228/129547192-b41d758f-e705-4bb4-b50d-6c6c3f3cf761.png)

I don't know how to modify the type definition of state

**Reply by tianzhich** [reply]
Well, I know your question. It's the limitation of the TS. You could add `?` mark after the `rootState` to eliminate this. For example:

```ts

const count = createModel<RootModel>()({
	state: 0, // initial state
	effects: () => ({
		incrementEffect(payload?: number, rootState?): number | undefined {
			if (rootState.count) {
				// do nothing
			}
			return payload
		},
	}),
})
```

**Reply by semoal** [reply]
Also you could convert id parameter to an object instead containing an optional Id

That will make the object required but the content of the object not.

You'll just need to check payload.id

**Reply by StringKe** [reply]
> Also you could convert id parameter to an object instead containing an optional Id
> 
> That will make the object required but the content of the object not.
> 
> You'll just need to check payload.id

thank you

**Reply by StringKe** [reply]
> Well, I know your question. It's the limitation of the TS. You could add `?` mark after the `rootState` to eliminate this. For example:
> 
> ```ts
> const count = createModel<RootModel>()({
> 	state: 0, // initial state
> 	effects: () => ({
> 		incrementEffect(payload?: number, rootState?): number | undefined {
> 			if (rootState.count) {
> 				// do nothing
> 			}
> 			return payload
> 		},
> 	}),
> })
> ```

thank you

This works, is this a feature of typescript? Does it have a corresponding document?

**Reply by semoal** [reply]
https://stackoverflow.com/questions/46958782/why-does-typescript-require-optional-parameters-after-required-parameters/46959804

This is a design limitation not by TypeScript itself, mostly JavaScript. You can read more about this here.
